test_opensearchpy/test_async/test_signer.py

```diff
@@ -103,3 +103,48 @@ async def test_aws_signer_async_frozen_credentials_as_http_auth(self) -> None:
         assert "X-Amz-Date" in headers
         assert "X-Amz-Security-Token" in headers
         assert len(mock_session.get_frozen_credentials.mock_calls) == 1
+
+
+class TestAsyncSignerWithSpecialCharacters:
+    def mock_session(self) -> Mock:
+        access_key = uuid.uuid4().hex
+        secret_key = uuid.uuid4().hex
+        token = uuid.uuid4().hex
+        dummy_session = Mock()
+        dummy_session.access_key = access_key
+        dummy_session.secret_key = secret_key
+        dummy_session.token = token
+
+        del dummy_session.get_frozen_credentials
+
+        return dummy_session
+    
+    async def test_aws_signer_async_with_special_chars_in_id(self) -> None:
+        region = "us-west-2"
+        
+        from opensearchpy.helpers.asyncsigner import AWSV4SignerAsyncAuth
+        from urllib.parse import quote
+
+        auth = AWSV4SignerAsyncAuth(self.mock_session(), region)
+
+        # ID contains special character
+        doc_id = "0x4786a3444661f1b9:0x625b2c53dfc9caba"
+        
+        # Test with unencoded ID
+        url_raw = f"https://search-domain.region.es.amazonaws.com/index/_doc/{doc_id}"
+        headers_raw = auth("GET", url_raw)
+        
+        # Test with properly encoded ID
+        encoded_id = quote(doc_id, safe='')
+        url_encoded = f"https://search-domain.region.es.amazonaws.com/index/_doc/{encoded_id}"
+        headers_encoded = auth("GET", url_encoded)
+        
+        # Signatures should be the same since URL encoding happens before signing
```

Comment:
Isn't it that currently the URL we send is not the URL we use for signing?

If so, we don't want the signing code to be decoding previously encoded URLs or knowing anything about URLs. The correct behavior would be that the URL passed into the client for an `index/_doc/{doc_id}` path is also the url that is used for signing. If we do any URL encoding, it should happen before signing, and then the HTTP request to the server uses the encoded version, too (so that the server can check it).

Reply:
Hmm that makes sense, I'll fix the test.